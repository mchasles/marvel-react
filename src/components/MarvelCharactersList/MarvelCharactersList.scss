.marvel-characters-list {
  background-color: #222;
  box-shadow: 0 -6px 6px #000;
  display: flex;
  flex-direction: column;
  height: 100vh;
  margin: 0;
  max-width: 360px;
  padding: 0;
  position: relative;
  width: 30%;

  @import 'Pagination';

  .react-logo-loader {
    width: 60px;
    height: 60px;
  }

  ul {
    margin: 0;
    padding: 0;
  }

  li {
    list-style-type: none;
    opacity: 0 ;
    animation: list-elem-fade-in 1 1s ease-in-out forwards;
    
    a {
      color: #999;
      display: block;
      padding: 10px;
      text-decoration: none;
    }

    a:hover, a.selected {
      background-color:rgba(38%, 85.5%, 98.4%, 0.7);
      color: white;
    }
  }

  $elements: 10;
  @for $i from 0 to $elements {
    li:nth-child(#{$i}) {
      animation-delay: #{$i * 10}ms;
    }
  }

  @keyframes list-elem-fade-in {
    from { opacity: 0 ; }
    to { opacity: 1 ; }
  }

  .error {
    color: #61dafb;
    font-size: small;
    font-style: italic;
    opacity: 0.5;
    position: relative;
    text-align: center;
    top: 50%;
    transform: translate(0, -50%);

    p {
      margin: 0;
    }
  }
}