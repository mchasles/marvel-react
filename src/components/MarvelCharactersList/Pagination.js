import React from 'react';
import arrowLeft from '../../assets/arrow-left.svg';
import arrowRight from '../../assets/arrow-right.svg';

export const Pagination = (props) => {
  return (
    <div className="pagination">
      <a className={(props.loading || props.offset === 0) ? 'disabled' : ''} onClick={!props.loading && props.previousCharacters}>
        <img src={arrowLeft} alt="Next" />
      </a>
      <a className={props.loading ? 'disabled' : ''} onClick={!props.loading && props.nextCharacters}>
        <img src={arrowRight} alt="Next" />
      </a>
    </div>
  );
}