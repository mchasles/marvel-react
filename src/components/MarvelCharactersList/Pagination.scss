  .pagination {
    display: flex;
    a {
      text-align: center;
      flex: 1;
      padding: 10px;
      
      &:hover {
        background-color:rgba(38%, 85.5%, 98.4%, 0.7);
      }
      
      &.disabled {
        opacity: 0.1;
        pointer-events: none;
      }
    }
  }