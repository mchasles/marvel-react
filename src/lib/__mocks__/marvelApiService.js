export const characters = [
  {
    'id': '1',
    'name': '3-D Man',
    'thumbnail': {
      'path': 'http://mock/1'
    }
  },
  {
    'id': '2',
    'name': 'A-Bomb (HAS)',
    'thumbnail': {
      'path': 'http://mock/2'
    }
  }
];

export const loadCharacters = () => {
  return new Promise((resolve, reject) => {
    resolve(characters);
  });
}

export const loadCharacter = (id) => {
  return new Promise((resolve, reject) => {
    let character = characters.find(ch => ch.id === id);
    character ? resolve(character) : reject({
      error: `Character with ${id} not found.`,
    })
  });
}

export const getCharacterPortraitSrc = (path) => {
  return new Promise((resolve, reject) => {
    if (path.length) {
      resolve(path + '/portrait.jpg');
    } else {
      reject('http://mock/default.jpg');
    }
  });
}