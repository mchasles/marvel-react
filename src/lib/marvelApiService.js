import axios from 'axios';
// API endpoint
const BASE_URL = 'https://gateway.marvel.com/v1/public/characters';
// API key
const API_KEY = 'cefd21ddeeea8bc5232fa3fe67c57e0f';
// URL of the default image when character portrait is not available
export const IMG_NOT_AVAILABLE_SRC = 'http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available/detail.jpg';
// URL suffix to add to resourceURI to get portrait
export const URL_PORTRAIT_SUFFIX  = '/detail.jpg';

// Fetch the 16 first characters from Marvel API (pagination will come later...)
export const loadCharacters = () => {
  return axios.get(BASE_URL, {
    params: {
      apikey: API_KEY,
      limit: 16,
      offset: 0
    }
  }).then(res => res.data.data.results)
};

// Fetch character info from Marvel API with its id
export const loadCharacter = (id) => {
  return axios.get(`${BASE_URL}/${id}`,{
    params: {
      apikey: API_KEY
    }
  }).then(res => res.data.data.results[0])
};

// Request character portrait image to check if available, if an error happens return default image src
export const getCharacterPortraitSrc = (path) => {
    let portraitSrc = `${path}${URL_PORTRAIT_SUFFIX}`;
    return axios.get(portraitSrc)
      .then(
        // Successfully get character portrait (but image might be the default one when not available)
        () => portraitSrc,
        // Error getting character portrait (error 404)
        () => IMG_NOT_AVAILABLE_SRC);
};