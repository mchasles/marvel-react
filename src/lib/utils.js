// Preload img in browser cache
export const preloadImg = (src) => {
  return new Promise((resolve, reject) => {
    const img = document.createElement('img');
    img.src = src;
    img.onload = () => resolve('done');
    // If image does not load after 6 seconds, reject promise
    setTimeout(() => {
      reject('error')
    }, 6000);
  });
}