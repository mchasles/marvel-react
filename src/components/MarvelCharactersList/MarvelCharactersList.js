import React from 'react';
import {Component} from 'react';
import {NavLink} from 'react-router-dom';

import {ReactLogoLoader} from '../ReactLogoLoader/';
import {Pagination} from './Pagination';

import './MarvelCharactersList.scss';

import {PAGE_LIMIT} from '../../lib/marvelApiService';

export class MarvelCharactersList extends Component {

  state = {
    offset: 0
  };

  nextCharacters() {
    let offset = this.state.offset + PAGE_LIMIT;
    this.props.updateCharacters(offset).then(() => {
      this.setState({offset});
    });
  }

  previousCharacters() {
    let offset = this.state.offset - PAGE_LIMIT;
    if (offset < 0) return;
    this.props.updateCharacters(offset).then(() => {
      this.setState({offset});
    });
  }

  render(){
    let content;
    if (this.props.error) {
      content = (
        <div className="error">
          <p>Error: Marvel API is unreachable</p>
          <p>Please retry later.</p>
        </div>
      );
    } else if (this.props.loading) {
      content = <ReactLogoLoader />;
    } else {
      content = (
        <div>
          <ul>
            {this.props.characters.map((ch) => (
                <li key={ch.id}>
                  <NavLink activeClassName="selected" to={`/character/${ch.id}`}>{ch.name}</NavLink>
                </li>
              ))}
          </ul>
        </div>
      );
    }
    return (
      <div className="marvel-characters-list">
        <Pagination 
          loading={this.props.loading} 
          offset={this.state.offset}
          previousCharacters={this.previousCharacters.bind(this)}
          nextCharacters={this.nextCharacters.bind(this)}
          />
        {content}
      </div>
    );
  }
}