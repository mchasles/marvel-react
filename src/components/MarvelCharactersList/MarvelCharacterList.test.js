import React from 'react';
import ReactDOM from 'react-dom';

import {MarvelCharactersList} from './MarvelCharactersList';

import renderer from 'react-test-renderer';
jest.mock('../../lib/marvelApiService');
import {characters} from '../../lib/marvelApiService';

jest.mock('react-router');

it('renders MarvelCharactersList without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<MarvelCharactersList characters={characters}/>, div);
});

it('renders MarvelCharactersList correctly displaying characters list loader', () => {
  const props = {
    error: false,
    loading: true,
    characters: null
  };
  const listLoading = renderer.create(React.createElement(MarvelCharactersList, props));
  expect(listLoading.toJSON()).toMatchSnapshot();
});

it('renders MarvelCharactersList correctly when characters are loaded', () => {
  const props = {
    error: false,
    loading: false,
    characters
  };
  const listLoaded = renderer.create(React.createElement(MarvelCharactersList, props));
  // TODO: check if there is a way to rend NavLink in snapshot
  expect(listLoaded.toJSON()).toMatchSnapshot();
});

it('renders MarvelCharactersList when error message when API does not respond', () => {
  const props = {
    error: true,
    loading: false,
    characters: null
  };
  const listLoadingFailed = renderer.create(React.createElement(MarvelCharactersList, props));
  expect(listLoadingFailed.toJSON()).toMatchSnapshot();
});