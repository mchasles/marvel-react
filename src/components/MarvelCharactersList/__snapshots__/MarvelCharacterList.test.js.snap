exports[`test renders MarvelCharactersList correctly displaying characters list loader 1`] = `
<div
  className="marvel-characters-list">
  <div
    className="pagination">
    <a
      className="disabled"
      onClick={false}>
      <img
        alt="Next"
        src="arrow-left.svg" />
    </a>
    <a
      className="disabled"
      onClick={false}>
      <img
        alt="Next"
        src="arrow-right.svg" />
    </a>
  </div>
  <img
    alt="React logo spinning"
    className="react-logo-loader"
    src="logo.svg" />
</div>
`;

exports[`test renders MarvelCharactersList correctly when characters are loaded 1`] = `
<div
  className="marvel-characters-list">
  <div
    className="pagination">
    <a
      className="disabled"
      onClick={[Function]}>
      <img
        alt="Next"
        src="arrow-left.svg" />
    </a>
    <a
      className=""
      onClick={[Function]}>
      <img
        alt="Next"
        src="arrow-right.svg" />
    </a>
  </div>
  <div>
    <ul>
      <li />
      <li />
    </ul>
  </div>
</div>
`;

exports[`test renders MarvelCharactersList when error message when API does not respond 1`] = `
<div
  className="marvel-characters-list">
  <div
    className="pagination">
    <a
      className="disabled"
      onClick={[Function]}>
      <img
        alt="Next"
        src="arrow-left.svg" />
    </a>
    <a
      className=""
      onClick={[Function]}>
      <img
        alt="Next"
        src="arrow-right.svg" />
    </a>
  </div>
  <div
    className="error">
    <p>
      Error: Marvel API is unreachable
    </p>
    <p>
      Please retry later.
    </p>
  </div>
</div>
`;
